lib/jxl/decode.cc

```diff
@@ -285,6 +285,72 @@ struct Sections {
   std::vector<char> section_received;
 };
 
+/*
+Given list of frame references to storage slots, and storage slots in which this
+frame is saved, computes which frames are required to decode the frame at the
+given index and any frames after it. The frames on which this depends are
+returned as a vector of their indices, in no particular order. The given index
+must be smaller than saved_as.size(), and references.size() must equal
+saved_as.size(). Any frames beyond saved_as and references are considered
+unknonw future frames and must be treated as if something depends on them.
+*/
+std::vector<size_t> GetFrameDependencies(size_t index,
+                                         const std::vector<int>& saved_as,
+                                         const std::vector<int>& references) {
```

Comment:
Shouldn't this also take a parameter that specifies if the frames here are *all* the frames?

Reply:
For now that's not implemented, as implemented now, it requires that all 8 referenceable storage slots are available after the frame with `index` and does not take the structure after that frame into account

This code could be upgraded with that, it would require adding everything from saved_as/references after index to the stack, and then depending on whether "seen_last" is true or not, require the 8 referenceable save slots or not as well. That is ok to do, but wouldn't do *that* much in the grand scheme of things given what the jxl encoder now does (which is to use slot 2 only, not taking dc into account).

So for now, I'd keep as is, and then taking the frames after index into account can be done as a future optimization.

Reply:
Fair enough, but shouldn't you push those slots in the stack too? I don't see that happening...

Reply:
It happens in the code pasted below. It goes through the 8 storage slots, and does not do 'continue' for all 8 of them when frame_index == index. Does this look correct?

```
      // Include all storage slots this frame depends on, except for the
      // future frames, which may depend on anything, so for index itself
      // test all 8 storage slots. We know some information about future frames
      // (up to the length of saved_as/references), but there can come more
      // frames after it.
      if (!(references[frame_index] & mask) && frame_index != index) continue;
```

Reply:
I separated out that code into explicitely pushing these values on the stack and added a TODO for the potential optimzation with all frames seen flag, PTAL